expression/builtin_op.go

```diff
@@ -81,14 +88,35 @@ func (b *builtinLogicAndSig) Clone() builtinFunc {
 	return newSig
 }
 
+func (b *builtinLogicAndSig) evalReal(row chunk.Row) (float64, bool, error) {
+	arg, isNull, err := b.evalInt(row)
+	return float64(arg), isNull, err
+}
+
 func (b *builtinLogicAndSig) evalInt(row chunk.Row) (int64, bool, error) {
-	arg0, isNull0, err := b.args[0].EvalInt(b.ctx, row)
-	if err != nil || (!isNull0 && arg0 == 0) {
-		return 0, err != nil, err
+	arg0, isNull0, err := b.args[0].EvalReal(b.ctx, row)
+	if err == ErrEvalRealCalled {
```

Comment:
I think this way to fix this problem is too crude, but I have no better idea now 😥.
I think that introducing a new method `evalBool` to `builtinFunc` is the best way, but it's not an easy way to add `evalBool` for all builtin functions.
@XuHuaiyu Do you have some better ideas?

Reply:
I think `evalReal` is just ok.
Mysql will convert string (or other type) to double,which means "1e-1000"(underflow) is false but "123qwe" is true.
Introducing `evalBool` will make too many change.

Reply:
It seems that we can add an `EvalBool` function for baseBuiltinFunc and using `switch-case` like [this](https://github.com/pingcap/tidb/blob/master/expression/scalar_function.go#L201-L235).
Then we can call EvalBool in LogicalAnd/LogicalOr/LogicalXor/IsTrue/IsFalse.EvalInt, etc.

What do you think? @qw4990 @zz-jason

Reply:
I agree with you.

Reply:
@qw4990 @XuHuaiyu I have implement the `evalReal` method. I think it will not be worse.If we met some similar bugs, we just need to fix `evalReal` methon. In most cases,`evalReal` should equal `evalInt` .